### PR TITLE
ENH: add state_changed callback on channels

### PR DESCRIPTION
**Pull request by klauer**
This may not be how it was intended to be used - here for discussion.
Should instead the virtual circuit be monitoring what happens to the
channels instead of the channels themselves?

### CODE REVIEW

File: caproto/_state.py [hunk 1]
```diff
@@ -188,10 +200,13 @@ def _fire_state_triggered_transitions(self, role):
...
             self.states[role], self.circuit_state.states[role] = new
+            return new
 
     def process_command(self, role, command_type):
-        self._fire_command_triggered_transitions(role, command_type)
-        self._fire_state_triggered_transitions(role)
+        command_trig = self._fire_command_triggered_transitions(role,
+                                                                command_type)
+        state_trig = self._fire_state_triggered_transitions(role)
+        return command_trig or state_trig
```
Comment (danielballan) [inline caproto/_state.py]: What if a command triggers and transition (`command_trig` is not None) and then a state transition is triggered (`state_trig` is not None). Don't we want `state_trig` returned in that case?

File: caproto/_hub.py [hunk 2]
```diff
@@ -586,6 +586,15 @@ def _fill_defaults(self, data_type, data_count):
             data_count = self.native_data_count
         return data_type, data_count
 
+    def state_changed(self, new_state):
```
Comment (danielballan) [inline caproto/_hub.py]: Is this just here in case a subclass wants to use it?

File: caproto/_state.py (hunk 1 above)
Comment (klauer) [inline caproto/_state.py]: You're right - swapped

File: caproto/_hub.py (hunk 2 above)
Comment (klauer) [inline caproto/_hub.py]: Yes - modified docstring as note

Comment (danielballan) [inline caproto/_hub.py]: It might be useful to have the old state too. I usually see this done as `self, new_state, old_state`.

**Comment by klauer** [comment]
One more attempt... I'm a bit rusty

**Comment by danielballan** [comment]
Even Rusty Ken keeps me on my toes.

**Pull request merged**